sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,187 @@
+// Package queue  provides Vault plugins with a Priority Queue. It can be used
+// as an in-memory list of queue.Item sorted by their priority, and offers
+// methods to find or remove items by their key. Internally it uses
+// container/heap; see Example Priority Queue:
+// https://golang.org/pkg/container/heap/#example__priorityQueue
+package queue
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+
+	"github.com/mitchellh/copystructure"
+)
+
+// ErrEmpty is returned for queues with no items
+var ErrEmpty = errors.New("queue is empty")
+
+// ErrDuplicateItem is returned when the queue attmepts to push an item to a key that
+// already exists. The queue does not attempt to update, instead returns this
+// error. If an Item needs to be updated or replaced, pop the item first.
+var ErrDuplicateItem = errors.New("duplicate item")
+
+// New initializes the internal data structures and returns a new
+// PriorityQueue
+func New() *PriorityQueue {
+	pq := PriorityQueue{
+		data:    make(queue, 0),
+		dataMap: make(map[string]*Item),
+	}
+	heap.Init(&pq.data)
+	return &pq
+}
+
+// PriorityQueue facilitates queue of Items, providing Push, Pop, and
+// PopByKey convenience methods. The ordering (priority) is an int64 value
+// with the smallest value is the highest priority. PriorityQueue maintains both
+// an internal slice for the queue as well as a map of the same items with their
+// keys as the index. This enables users to find specific items by key. The map
+// must be kept in sync with the data slice.
+// See https://golang.org/pkg/container/heap/#example__priorityQueue
+type PriorityQueue struct {
+	// data is the internal structure that holds the queue, and is operated on by
+	// heap functions
+	data queue
```

Comment:
Should this be a pointer to the queue instead of the queue object?  I see other places where you just pass the reference and that could just be stored here.

Reply:
So that I understand what you're asking, are the other places heap methods like `heap.Init(&pq.data)` and `heap.Pop(&pq.data).(*Item)`, and you're asking if we make `data` a `*queue` instead of `queue`, so those calls are `heap.Init(pq.data)` et. al? 

Assuming that's what you're asking,, is there an advantage to doing that? I'm not opposed to it I'm just curious of the benefits.

Reply:
I'm fine either way, it just seems that you are always using the reference to it in all calls instead of just storing the reference and just using it directly.  It is really just a stylistic thing.  If I see a bunch of `&` for the same thing, I start digging into the implementation.